src/apps/SealedBidTokenSale.sol

```diff
@@ -0,0 +1,270 @@
+     * @dev - Sale must be ended successfully and Merkle root set
+     *      - User must not have claimed already
+     *      - Proof must be valid for user's allocation
+     *      - Transfers allocated tokens and USDC to user
+     * @param saleTokenAllocation Amount of sale tokens allocated to user
+     * @param usdcAllocation Amount of USDC allocated to user
+     * @param proof Merkle proof verifying the allocation
+     * @param user Address of user claiming tokens
+     */
+    function claimTokens(uint256 saleTokenAllocation, uint256 usdcAllocation, bytes32[] calldata proof, address user)
+        external
+        nonReentrant
+    {
+        // Verify sale ended successfully and claims are enabled
+        if (!saleEnded || !capReached) revert CapNotReached();
+        if (merkleRoot == bytes32(0)) revert MerkleRootNotSet();
+        if (hasClaimed[user]) revert AlreadyClaimed(user);
+
+        // Create and verify Merkle leaf
+        bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(user, saleTokenAllocation, usdcAllocation))));
+        if (!MerkleProof.verify(proof, merkleRoot, leaf)) revert InvalidProof(proof, leaf);
+
+        // Mark as claimed before transfers
+        hasClaimed[user] = true;
+
+        // Transfer allocated sale tokens if any
+        if (saleTokenAllocation > 0) {
+            saleToken.safeTransfer(user, saleTokenAllocation);
+        }
+
+        // Transfer allocated USDC if any
+        if (usdcAllocation > 0) {
+            USDC.safeTransfer(user, usdcAllocation);
+        }
+
+        // Emit claim event
+        emit Claimed(user, saleTokenAllocation);
+    }
+
+    /* ============ Admin Functions ============ */
+
+    /**
+     * @notice Allows owner to officially end the sale
+     * @dev - Can only be called once
+     *      - Sets final sale status based on minimum cap
+     *      - Emits event with final sale results
+     */
+    function endSale() external onlyOwner {
+        // Verify sale hasn't already been ended
+        if (saleEnded) revert SaleAlreadyEnded(block.timestamp);
+
+        // Mark sale as ended and determine if cap was reached
+        saleEnded = true;
+        capReached = totalDeposited >= minimumCap;
+
+        // Emit sale end event with final status
+        emit SaleEnded(capReached, totalDeposited);
+    }
+
+    /**
+     * @notice Sets the Merkle root for verifying token allocations
+     * @dev - Sale must be ended successfully
+     *      - Enables token claiming process
+     * @param newRoot The Merkle root hash of all valid allocations
+     */
+    function setMerkleRoot(bytes32 newRoot) external onlyOwner {
+        // Verify sale ended successfully before setting root
+        if (!saleEnded || !capReached) revert CapNotReached();
+
+        // Update Merkle root
+        merkleRoot = newRoot;
+
+        // Emit root update event
+        emit MerkleRootSet(newRoot);
+    }
+
```

Comment:
Missing a view function that given an address returns the following

function saleStatus(address user) returns SaleInfo

startTime
minimumCap
totalDeposits
saleEnded
capReached
hasClaimed[user]
deposits[user]